cosmic-applet-battery/src/power_daemon.rs

```diff
@@ -64,4 +85,135 @@ trait PowerDaemon {
+    match power.as_str() {
+        "Battery" => Ok(Power::Battery),
+        "Balanced" => Ok(Power::Balanced),
+        "Performance" => Ok(Power::Performance),
+        _ => panic!("Unknown power profile: {}", power),
+    }
+}
+
+pub async fn set_power_profile(daemon: PowerDaemonProxy<'_>, power: Power) -> Result<()> {
+    match power {
+        Power::Battery => daemon.battery().await,
+        Power::Balanced => daemon.balanced().await,
+        Power::Performance => daemon.performance().await,
+    }
+}
+
+pub fn power_profile_subscription<I: 'static + Hash + Copy + Send + Sync + Debug>(
+    id: I,
+) -> iced::Subscription<(I, PowerProfileUpdate)> {
+    subscription::unfold(id, State::Ready, move |state| start_listening(id, state))
+}
+
+#[derive(Debug)]
+pub enum State {
+    Ready,
+    Waiting(Connection, UnboundedReceiver<PowerProfileRequest>),
+    Finished,
+}
+
+async fn start_listening<I: Copy>(id: I, state: State) -> (Option<(I, PowerProfileUpdate)>, State) {
+    match state {
+        State::Ready => {
+            let conn = match Connection::system().await.map_err(|e| e.to_string()) {
+                Ok(conn) => conn,
+                Err(e) => return (Some((id, PowerProfileUpdate::Error(e))), State::Finished),
+            };
+            let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+            let power_proxy = match PowerDaemonProxy::new(&conn)
+                .await
+                .map_err(|e| e.to_string())
+            {
+                Ok(p) => p,
+                Err(e) => {
+                    return (
+                        Some((id, PowerProfileUpdate::Error(e))),
+                        State::Waiting(conn, rx),
+                    )
+                }
+            };
+            let profile = match get_power_profile(power_proxy)
+                .await
+                .map_err(|e| e.to_string())
+            {
+                Ok(p) => p,
+                Err(e) => {
+                    return (
+                        Some((id, PowerProfileUpdate::Error(e))),
+                        State::Waiting(conn, rx),
+                    )
+                }
+            };
+
+            return (
+                Some((id, PowerProfileUpdate::Init(profile, tx))),
+                State::Waiting(conn, rx),
+            );
+        }
+        State::Waiting(conn, mut rx) => {
+            let power_proxy = match PowerDaemonProxy::new(&conn)
+                .await
+                .map_err(|e| e.to_string())
+            {
+                Ok(p) => p,
+                Err(e) => {
+                    return (
+                        Some((id, PowerProfileUpdate::Error(e))),
+                        State::Waiting(conn, rx),
+                    )
+                }
+            };
+
+            match rx.recv().await {
+                Some(PowerProfileRequest::Get) => {
+                    if let Ok(profile) = get_power_profile(power_proxy).await {
+                        return (
```

Comment:
These returns are at the end of the expression, so the `return` keyword is unneeded. You should be able to just remove `return` and the final `;`.